src/Week4/HW4.hs

```diff
@@ -0,0 +1,133 @@
+showTuple :: (Num a, Eq a, Show a) => [(a,a)] -> String
+showTuple []     = ""
+showTuple ((a,ord):as) = case ord of
+  0 -> show a ++ rest
+  1 -> if a == 1          then "x" ++ rest 
+        else if a == (-1) then "-x" ++ rest
+         else show a ++ "x" ++ rest
+  _ -> if a == 1          then "x^" ++ show ord ++ rest
+        else if a == (-1) then "-x^" ++ show ord ++ rest 
+         else show a ++ "x^" ++ show ord ++ rest
+  where
+   rest = if length as > 0 then " + " ++ showTuple as else ""
+        
+-- Exercise 4 -----------------------------------------
+
+-- define addition for the Poly type
+plus :: Num a => Poly a -> Poly a -> Poly a
+plus (P a) (P b) = P (addLists a b) 
+
+-- zipWith (+) truncates the sum to match the size of the shortest of the two input lists, which is not what we want
+-- so I defined a new function
+addLists :: (Num a) => [a] -> [a] -> [a]
+addLists (a:as) (b:bs) = (a+b) : addLists as bs
+addLists [] b          = b
+addLists a []          = a
+
+-- Exercise 5 -----------------------------------------
+
+-- implement polynomial multiplication
+times :: Num a => Poly a -> Poly a -> Poly a
+times (P a) (P b) = P (timesLists a b)
+
+-- helpers:
+-- given lists a and b and an index, multiply every item of b by a!!index (and add leading zeroes accordingly)
+timesOneCoeff :: (Num a) => [a] -> [a] -> Int -> [a]
+timesOneCoeff a b num     = map (\j -> j * (a!!num)) (replicate num 0 ++ b) 
+
+-- multiply two polynomial lists by executing timesOneCoeff for each index of a, and summing the subresults together
+timesLists :: (Num a) => [a] -> [a] -> [a]
+timesLists a b = helper a b (length a)
+  where helper _ _ 0   = []
+        helper l1 l2 num = addLists (timesOneCoeff l1 l2 (num-1)) (helper l1 l2 (num-1))
+
+-- Exercise 6 -----------------------------------------
+
+-- define negate, which negates the input polynomial, and fromInteger, which takes an integer a and returns Poly a
+instance Num a => Num (Poly a) where
+    (+) = plus
+    (*) = times
+    negate (P a)  = P $ Prelude.negate <$> a
+    fromInteger n = P [Prelude.fromInteger n]
+    -- No meaningful definitions exist
+    abs    = undefined
+    signum = undefined
+
+-- Exercise 7 -----------------------------------------
+
+-- apply a polynomial to a value
+applyP :: (Num a, Integral a) => Poly a -> a -> a
+applyP (P a) n = sum $ (<$>) (\(coeff, ord) -> coeff * n^ord) $ toTuple a
+
+-- Exercise 8 -----------------------------------------
+
+--implement the nderiv (nth derivative) function
+class Num a => Differentiable a where
+    deriv  :: a -> a
+    nderiv :: Int -> a -> a
+    nderiv n = case n of
+                1 -> deriv
+                _ -> deriv . nderiv (n-1)
+
+-- Exercise 9 -----------------------------------------
+
+-- define a function for polynomial differentiation
+instance (Num a, Eq a) => Differentiable (Poly a) where
+    deriv (P a) = P $ drop 1 $ map (\(coeff, _) -> coeff) $ map (\(coeff, ord) -> (coeff * ord, ord-1)) $ toTuple a
```

Comment:
No reason putting ord-1 as second element of the tuple and then immediately drop it in the next map.
```suggestion
    deriv (P a) = P $ drop 1 $ map (\(coeff, ord) -> coeff * ord) $ toTuple a
```